metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java

```diff
@@ -149,17 +151,53 @@ public String toString() {
 
     private final SnapshotRegistry snapshotRegistry;
     private final Logger log;
-    private final Random random;
+
+    /**
+     * The random number generator used by this object.
+     */
+    private final RandomSource random;
+
+    /**
+     * The KIP-464 default replication factor that is used if a CreateTopics request does
+     * not specify one.
+     */
     private final int defaultReplicationFactor;
+
+    /**
+     * The KIP-464 default number of partitions that is used if a CreateTopics request does
+     * not specify a number of partitions.
+     */
     private final int defaultNumPartitions;
+
+    /**
+     * A reference to the controller's configuration control manager.
+     */
     private final ConfigurationControlManager configurationControl;
-    private final ClusterControlManager clusterControl;
+
+    /**
+     * A reference to the controller's cluster control manager.
+     */
+    final ClusterControlManager clusterControl;
+
+    /**
+     * Maps topic names to topic UUIDs.
+     */
     private final TimelineHashMap<String, Uuid> topicsByName;
+
+    /**
+     * Maps topic UUIDs to structures containing topic information, including partitions.
+     */
     private final TimelineHashMap<Uuid, TopicControlInfo> topics;
 
+    /**
+     * A map of broker IDs to the partitions that the broker leads.
+     * Partitions with no leader appear in this map under id -1.
+     */
+    private final TimelineHashMap<Integer, TimelineHashMap<Uuid, int[]>> leaderships;
```

Comment:
Wondering if it's better to have a mapping from Broker ➝ All partitions on that broker and not just leaders. For example with controlled shutdown, we'd be generating partition records for not just leaders but also other partitions that need to have this broker removed from their isr/replica list. @cmccabe